### ISSUE TITLE
Proposed solution to the AR includes problem

### ISSUE THREAD

**Issue by gjb83** [alexreisner/geocoder#99]
I would like to propose the following solution to the AR includes problem (see 'Known Issue' at the end of the README), and to open a discussion with Alex and the geocoder community.

Proposed solution:
1. Remove the `:select` from [the AR query](https://github.com/alexreisner/geocoder/blob/master/lib/geocoder/stores/active_record.rb#L160-162).
2. Instead, add `distance` and `bearing` instance methods to the AR model, to be calculated application-side. The formulae would translate easily into pure Ruby code.

This would primarily enable the `includes` scope to work correctly, but would also improve performance:
- The bearing is currently always calculated, but may not be required.
- The database must currently calculate the distance twice per row: once for the `SELECT` and again for the `ORDER BY`.
- When used in conjunction with pagination, many of the calculated distances and bearings are discarded if they don't appear on the current page, only to be recalculated when the page changes.

Comments welcome.

**Comment by amnesia7** [comment]
Disregard my comment above. Although it allows me to include extra fields in the `.select()` to use for ordering the results they are not callable in the view so if distance was included in the select the view would not be able to output it.
You could try `.joins(:categories).preload(:categories)`
http://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-preload

**Comment by rubyconvict** [comment]
Thanks @amnesia7 , it works, though it sucks in terms of performance (N+1 problem).

**Comment by amnesia7** [comment]
@rubyconvict maybe try `.joins(:categories).eager_load(:categories)` instead.

**Comment by rubyconvict** [comment]
No, I tried, this one is not any better (N+1) and I have one search scenario where it causes "ERROR: PG::Error: ERROR: column "distance" does not exist".

**Comment by amnesia7** [comment]
Hmmm, I thought `.join` would allow you to link the tables whilst still allowing you to use your own `.select` and then `.eager_load` would mean that the other objects were loaded as well...I guess not then.

**Comment by rubyconvict** [comment]
I changed my mind: `.joins(:categories)` needs a `group` clause, otherwise results can be duplicated, but when I do `group`, I see the same error, so for now, best solution for me is to `.preload(:categories)` and use `.sort_by{ |i| i.name }` on each categories array.

**Comment by brandoncc** [comment]
@valscion was [your solution](https://github.com/alexreisner/geocoder/issues/99#issuecomment-35388931) ever proven to work? @alexreisner this solution looks the most promising out of all of the threads/issues I have read about this, any luck with it?

**Comment by valscion** [comment]
@brandoncc I no longer have this issue as we've changed our search query building significantly since I posted my comment. Feel free to try the code out, but I'm no longer sure whether it works or not. YMMV

**Comment by brandoncc** [comment]
@valscion Thanks for responding. I ended up forking and trying your code. It didn't work, I'm not sure if that is because I am using Rails 4 or not. Are you still searching and ordering by distance? Did you figure out another way using Geocoder that works properly and might help the rest of us? Thanks again.

**Comment by valscion** [comment]
@brandoncc No, I'm no longer ordering by distance, although we're searching by it still.
